pkg/gce-pd-csi-driver/controller_test.go

```diff
@@ -68,8 +68,248 @@ var (
+		},
+		{
+			name: "fail no source volume found",
+			req: &csi.CreateSnapshotRequest{
+				Name:           name,
+				SourceVolumeId: "/test/wrongname",
+			},
+			expErrCode: codes.NotFound,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Logf("test case: %s", tc.name)
+		// Setup new driver each time so no interference
+		gceDriver := GetGCEDriver()
+		fakeCloudProvider, err := gce.FakeCreateCloudProvider(project, zone)
+		if err != nil {
+			t.Fatalf("Failed to create fake cloud provider: %v", err)
+		}
+		err = gceDriver.SetupGCEDriver(fakeCloudProvider, nil, nil, metadataservice.NewFakeService(), driver, "vendor-version")
+		if err != nil {
+			t.Fatalf("Failed to setup GCE Driver: %v", err)
+		}
+
+		// Start Test
+		resp, err := gceDriver.cs.CreateSnapshot(context.Background(), tc.req)
+		//check response
+		if err != nil {
+			serverError, ok := status.FromError(err)
+			if !ok {
+				t.Fatalf("Could not get error status code from err: %v", serverError)
+			}
+			if serverError.Code() != tc.expErrCode {
+				t.Fatalf("Expected error code: %v, got: %v. err : %v", tc.expErrCode, serverError.Code(), err)
+			}
+			continue
+		}
+		if tc.expErrCode != codes.OK {
+			t.Fatalf("Expected error: %v, got no error", tc.expErrCode)
+		}
+
+		// Make sure responses match
+		snapshot := resp.GetSnapshot()
+		if snapshot == nil {
+			// If one is nil but not both
+			t.Fatalf("Expected snapshot %v, got nil snapshot", tc.expSnapshot)
+		}
+
+		if !reflect.DeepEqual(snapshot, tc.expSnapshot) {
+			errStr := fmt.Sprintf("Expected snapshot: %#v\n to equal snapshot: %#v\n", snapshot, tc.expSnapshot)
+			t.Errorf(errStr)
+		}
+	}
+}
+func TestDeleteSnapshot(t *testing.T) {
+	testCases := []struct {
+		name       string
+		req        *csi.DeleteSnapshotRequest
+		expErrCode codes.Code
+	}{
+		{
+			name: "valid",
+			req: &csi.DeleteSnapshotRequest{
+				SnapshotId: testSnapshotId,
+			},
+		},
+		{
+			name: "invalid id",
+			req: &csi.DeleteSnapshotRequest{
+				SnapshotId: testSnapshotId + "/foo",
+			},
+		},
+		{
+			name: "empty id",
+			req: &csi.DeleteSnapshotRequest{
+				SnapshotId: "",
+			},
+			expErrCode: codes.InvalidArgument,
+		},
+	}
+	for _, tc := range testCases {
+		t.Logf("test case: %s", tc.name)
+		// Setup new driver each time so no interference
+		gceDriver := initGCEDriver(t)
+
+		_, err := gceDriver.cs.DeleteSnapshot(context.Background(), tc.req)
+		//check response
+		if err != nil {
+			serverError, ok := status.FromError(err)
+			t.Logf("get server error %v", serverError)
+			if !ok {
+				t.Fatalf("Could not get error status code from err: %v", serverError)
+			}
+			if serverError.Code() != tc.expErrCode {
+				t.Fatalf("Expected error code: %v, got: %v. err : %v", tc.expErrCode, serverError.Code(), err)
+			}
+			continue
+		}
+		if tc.expErrCode != codes.OK {
+			t.Fatalf("Expected error: %v, got no error", tc.expErrCode)
+		}
+
+	}
+}
+
+func TestListSnapshotsArguments(t *testing.T) {
```

Comment:
what about tests where we dont specify the ID and need to list all snapshots

Reply:
The the last test "no id" is to return all the snapshots.